cirq/qis/states.py

```diff
@@ -35,6 +36,435 @@
     np.ndarray, Sequence[Union[int, float, complex]],
     # Product state object
     'cirq.ProductState',]
+document(
+    STATE_VECTOR_LIKE,  # type: ignore
+    """An object representing a state vector.""")
+
+QUANTUM_STATE_LIKE = Union[
+    # state vector
+    STATE_VECTOR_LIKE,
+    # density matrix
+    np.ndarray,
+    # quantum state object
+    'cirq.QuantumState']
+document(
+    QUANTUM_STATE_LIKE,  # type: ignore
+    """An object representing a quantum state.""")
+
+
+class QuantumState:
+    """A quantum state.
+
+    Can be a state vector, a state tensor, or a density matrix.
+    """
+
+    def __init__(
+            self,
+            data: np.ndarray,
+            qid_shape: Optional[Tuple[int, ...]] = None,
+            *,  # Force keyword arguments
+            dtype: Type[np.number] = np.complex64,
+            validate: bool = True,
+            atol: float = 1e-7) -> None:
+        """Initialize a quantum state object.
+
+        Args:
+            data: The data representing the quantum state.
+            qid_shape: The qid shape.
+            validate: Whether to check if the the given data and qid shape
+                represent a valid quantum state.
+            dtype: The data type.
+            atol: Absolute numerical tolerance to use for validation.
+
+        Raises:
+            ValueError: The qid shape was not specified and could not be
+                inferred.
+        """
+        if qid_shape is None:
+            qid_shape = infer_qid_shape(data)
+        self._data = data
+        self._qid_shape = qid_shape
+        self._dim = np.prod(self.qid_shape, dtype=int)
+        if validate:
+            self.validate(dtype=dtype, atol=atol)
+
+    @property
+    def data(self) -> np.ndarray:
+        return self._data
+
+    @property
+    def qid_shape(self) -> Tuple[int, ...]:
+        return self._qid_shape
+
+    @property
+    def dtype(self) -> np.ndarray:
+        return self._data.dtype
+
+    def is_state_vector(self) -> bool:
+        return self.data.shape == (self._dim,)
+
+    def is_state_tensor(self) -> bool:
+        return self.data.shape == self.qid_shape
+
+    def is_density_matrix(self) -> bool:
+        return self.data.shape == (self._dim, self._dim)
+
+    def is_pure_state(self) -> bool:
```

Comment:
I'm uneasy about this function because it may mislead the caller when `data` is a density matrix that correspond to a pure state such as [[1, 0], [0, 0]].

It seems to me that we don't actually need it: we could replace calls to `is_pure_state` with `not is_density_matrix`.

Reply:
Done.